src/lib.rs

```diff
@@ -0,0 +1,324 @@
+use std::borrow::Cow;
+
+use once_cell::sync::OnceCell;
+use eyre::ContextCompat;
+use serde::{Serialize, Deserialize};
+use serde_json::Value;
+use json_value_merge::Merge;
+use dyn_fmt::AsStrFormatExt;
+
+use fluvio_smartmodule::{
+    smartmodule, Result, SmartModuleRecord, RecordData,
+    dataplane::smartmodule::{
+        SmartModuleExtraParams, SmartModuleInitError
+    },
+    eyre
+};
+
+static PARAMS: OnceCell<Params> = OnceCell::new();
+const PARAM_NAME: &str = "spec";
+
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Params {
+    #[serde(rename="match")]
+    _match: Option<Vec<Match>>,
+    default: Default,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Match {
+    key: String,
+    value: String,
+    format: Format
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Default {
+    format: Format
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Format {
+    with: String,
+    using: Vec<String>,
+    output: String
+}
+
+/// Look-up values in json [ "/top/one", "/top/two"], and generate formatted string
+fn make_formatted_string(v: &Value, lookup: &Vec<String>, format_str: &String) -> Result<String> {
```

Comment:
Since it is converting to Cow later, maybe better to accept Cow as input. Not show stopper

Reply:
Yeah, I'm not comfortable enough with Cow. It's something @galibey helped me with on a previous SmartModule.